File: backend/api/ratelimit.py
"""Abuse control for the public science endpoints.

Every route below ``/api`` is unauthenticated, and most of them are cheap. The one that is not
cheap is the catalog: a request naming an unseen NORAD id reaches Celestrak, which is a shared
third-party service that degrades under bursts -- 500s and 503s have been observed from it. An
unthrottled public endpoint in front of it turns one scripted caller into this project's outage
*and* a nuisance to Celestrak.

So the limits here exist to protect an upstream dependency and this deployment's CPU, not a paid
quota. That is why there is no global budget as there is on the agent route: there is no shared
spend to exhaust, only a shared service to be rude to.

**Limits are deliberately generous.** The solar-system view fetches on a debounced time slider,
so a person scrubbing through dates legitimately produces a steady stream of requests. A limit
tight enough to feel like protection would break ordinary use, which is the failure mode that
gets a limiter deleted rather than tuned.

What this does not guarantee: state lives in this process's memory, so with several instances
each carries its own counters. The effective ceiling is ``instances x limit``. This narrows the
blast radius; it is not a hard cap. A hard cap needs shared state, which is a provisioned
dependency and a hosting decision.
"""

from __future__ import annotations

import os
import time
from collections import OrderedDict, deque
from dataclasses import dataclass

from fastapi import Request
from fastapi.responses import JSONResponse

# Paths that are never limited. Health is polled by monitoring, and a limiter that can make a
# service look unhealthy has made the problem it was meant to prevent.
EXEMPT_PATHS = frozenset({"/api/health"})


@dataclass(frozen=True)
class Limit:
    """One sliding window: how many requests, over how long."""

    max_requests: int
    window_s: float

    def describe(self) -> str:
        """Human-readable form of the limit, for the error a caller actually reads."""
        if self.window_s >= 3600:
            return f"{self.max_requests} requests per hour"
        if self.window_s >= 60:
            return f"{self.max_requests} requests per minute"
        return f"{self.max_requests} requests per {self.window_s:g} seconds"


# A person dragging the time slider produces a debounced but steady stream, so the per-minute
# figure is set well above interactive use and aimed at scripted traffic instead.
DEFAULT_LIMITS: tuple[Limit, ...] = (
    Limit(max_requests=120, window_s=60),
    Limit(max_requests=1200, window_s=3600),
)

MAX_TRACKED_CLIENTS = 10_000


@dataclass(frozen=True)
class Decision:
    """Outcome of a limit check."""

    allowed: bool
    message: str = ""
    retry_after_s: int = 0


class SlidingWindowLimiter:
    """Per-client sliding-window limiter holding its counters in process memory.

    Several windows are checked together, so a caller can be held to a burst rate and a sustained
    rate at once. Checking only a burst window lets a caller sustain that burst rate forever;
    checking only a long window lets it spend the whole hour's allowance in a second.
    """

    def __init__(
        self,
        limits: tuple[Limit, ...] = DEFAULT_LIMITS,
        max_tracked_clients: int = MAX_TRACKED_CLIENTS,
    ) -> None:
        """Build a limiter.

        Args:
            limits: Windows to enforce. The widest window decides how long history is kept.
            max_tracked_clients: Ceiling on remembered clients, so the map cannot grow unbounded.
        """
        self._limits = limits
        self._max_tracked_clients = max_tracked_clients
        self._widest_window_s = max(limit.window_s for limit in limits)
        # Ordered by recency of last request, so eviction can drop the stalest client in O(1).
        self._clients: OrderedDict[str, deque[float]] = OrderedDict()

    def check(self, client_id: str, now: float | None = None) -> Decision:
        """Decide whether a request may proceed, recording it if so.

        Args:
            client_id: Stable identifier for the caller.
            now: Current monotonic-ish time in seconds; injectable so the policy is testable
                without sleeping.

        Returns:
            Whether the request is allowed, and if not, why and for how long to wait.
        """
        moment = time.time() if now is None else now

        hits = self._clients.get(client_id)
        if hits is None:
            hits = deque()
        else:
            self._clients.move_to_end(client_id)

        cutoff = moment - self._widest_window_s
        while hits and hits[0] <= cutoff:
            hits.popleft()

        for limit in self._limits:
            window_start = moment - limit.window_s
            in_window = [at for at in hits if at > window_start]

            if len(in_window) >= limit.max_requests:
                # Store the pruned history even on refusal, so retrying through a refusal cannot
                # keep a caller's window from expiring.
                self._clients[client_id] = hits
                self._evict_if_crowded()
                oldest = in_window[0]
                return Decision(
                    allowed=False,
                    message=(
                        f"Too many requests. This service allows {limit.describe()} per client."
                    ),
                    retry_after_s=max(1, int(oldest + limit.window_s - moment) + 1),
                )

        hits.append(moment)
        self._clients[client_id] = hits
        self._clients.move_to_end(client_id)
        self._evict_if_crowded()

        return Decision(allowed=True)

    def tracked_clients(self) -> int:
        """How many clients are currently held in memory. Exposed so the bound is assertable."""
        return len(self._clients)

    def _evict_if_crowded(self) -> None:
        """Drop least-recently-seen clients so memory stays bounded.

        Expiry alone is not enough: a flood of distinct identifiers adds entries faster than the
        widest window retires them, and an unbounded map in a reused instance is a leak.
        """
        while len(self._clients) > self._max_tracked_clients:
            self._clients.popitem(last=False)


def client_id_from_request(request: Request) -> str:
    """Identify the caller from proxy headers.

    Vercel sets ``x-forwarded-for`` with the client address first. The header is caller-supplied
    in principle, so this is a cooperative identifier rather than proof of identity. It is the
    right granularity for protecting an upstream from accidental hammering, which is what this
    limiter is for; it is not an authentication mechanism and is not used as one.

    Args:
        request: Incoming request.

    Returns:
        A stable key for the caller, or a shared fallback when no address is present.
    """
    forwarded = request.headers.get("x-forwarded-for")
    if forwarded:
        first = forwarded.split(",")[0].strip()
        if first:
            return first

    real = (request.headers.get("x-real-ip") or "").strip()
    if real:
        return real

    if request.client and request.client.host:
        return request.client.host

    return "unidentified"


def rate_limiting_enabled() -> bool:
    """Whether limits apply in this environment.

    Enforced only on a deployment, detected by ``VERCEL_ENV``. The evaluation harness drives
    hundreds of requests through a local server in one run, and a limiter that throttled that
    would be measuring itself rather than the tools. The policy is still verified in
    development -- by its own tests, which drive the limiter directly.

    Returns:
        True when running on a Vercel deployment.
    """
    return bool(os.environ.get("VERCEL_ENV"))


limiter = SlidingWindowLimiter()


async def rate_limit_middleware(request: Request, call_next):  # type: ignore[no-untyped-def]
    """Refuse requests from a client that is over its allowance.

    Args:
        request: Incoming request.
        call_next: The rest of the application stack.

    Returns:
        The downstream response, or a 429 with a typed error body and ``Retry-After``.
    """
    if not rate_limiting_enabled() or request.url.path in EXEMPT_PATHS:
        return await call_next(request)

    decision = limiter.check(client_id_from_request(request))
    if decision.allowed:
        return await call_next(request)

    return JSONResponse(
        status_code=429,
        content={"error": {"code": "rate_limited", "message": decision.message}},
        headers={"Retry-After": str(decision.retry_after_s)},
    )

File: src/lib/agent/rateLimit.ts
/**
 * Abuse control for the public agent endpoint.
 *
 * `/api/agent` is unauthenticated by design -- a stranger has to be able to ask a question for
 * the project to make its argument at all. That openness spends a real provider quota, so the
 * endpoint needs a budget rather than a door.
 *
 * Two limits, because they defend against different things:
 *
 * - **Per client**, which stops one caller monopolising the service.
 * - **A global daily budget**, which stops the provider quota being drained at all. Per-client
 *   limits alone do not defend it: the quota is shared, so enough distinct callers exhaust it
 *   while every one of them stays politely inside its own allowance.
 *
 * The global budget is charged in **model calls, after the fact**, not in requests. A grounded
 * question makes one model call per tool-calling round, so counting requests would either
 * under-count real usage or have to assume the worst case on every question. The agent reports
 * how many steps it actually took; that is what gets charged.
 *
 * ## What this does not guarantee
 *
 * State lives in this process's memory. Vercel's Fluid Compute reuses instances, so a sustained
 * caller does keep hitting the same counters -- but under concurrency there can be several
 * instances, and each carries its own. The effective global ceiling is therefore
 * `instances x DAILY_MODEL_CALL_BUDGET`, not `DAILY_MODEL_CALL_BUDGET`.
 *
 * That is stated rather than papered over: this is a meaningful reduction in blast radius, not
 * a hard cap. A hard cap needs shared state (Redis or similar), which is a hosting decision and
 * a provisioned dependency. The honest version of this control is the one that says which of the
 * two it is.
 */

/** Tunables, kept together so the policy is readable in one place. */
export type RateLimitConfig = {
  /** Requests allowed per client inside `burstWindowMs`. */
  burstLimit: number;
  burstWindowMs: number;
  /** Requests allowed per client inside `sustainedWindowMs`. */
  sustainedLimit: number;
  sustainedWindowMs: number;
  /** Model calls allowed across all clients inside `budgetWindowMs`. */
  budgetLimit: number;
  budgetWindowMs: number;
  /** Ceiling on tracked clients, so the map cannot grow without bound. */
  maxTrackedClients: number;
};

const SECOND = 1000;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

/**
 * Default policy.
 *
 * The daily budget is set below the Gemini free tier's 500 requests per day for the lite model,
 * so that exhaustion produces *this* endpoint's clear 429 rather than the provider's opaque one.
 * A caller who sees "the service has used its budget for today" can act on it; a raw provider
 * quota error reads like a bug.
 *
 * **The figure is per instance, so it is chosen for the aggregate, not for one process.** The
 * budget cannot be shared without shared state, so under concurrency the real ceiling is
 * `instances x budgetLimit`. Setting this to the provider's own cap would therefore overshoot it
 * by whatever the instance count happens to be. At 150 the aggregate stays near 500 across the
 * handful of instances this workload actually spreads over, while a single-instance deployment
 * -- the common case -- still serves 150 questions a day, far more than a demo needs.
 *
 * The provider's free tier is itself a hard cap, and it is the one that bounds *spend*: with no
 * billing attached, exceeding it costs nothing and simply fails. This budget exists to make that
 * boundary legible and to keep one caller from consuming the day's allowance, not to be the last
 * line of defence against a bill. If a paid key is ever attached, that changes: the budget
 * becomes the only thing between a caller and real money, and a shared counter earns its keep.
 */
export const DEFAULT_RATE_LIMIT: RateLimitConfig = {
  burstLimit: 5,
  burstWindowMs: MINUTE,
  sustainedLimit: 40,
  sustainedWindowMs: HOUR,
  budgetLimit: 150,
  budgetWindowMs: DAY,
  maxTrackedClients: 10_000,
};

/** Why a request was refused, and how long the caller should wait. */
export type RateLimitDecision =
  | { allowed: true }
  | {
      allowed: false;
      code: "rate_limited" | "daily_budget_exhausted";
      message: string;
      retryAfterS: number;
    };

/** One client's request timestamps, newest last. */
type ClientRecord = {
  /** Millisecond timestamps of recent requests, ascending. */
  hits: number[];
};

/** A charge against the shared budget: when it happened, and how many model calls it cost. */
type BudgetEntry = {
  at: number;
  cost: number;
};

/**
 * Drop timestamps that have fallen out of a sliding window.
 *
 * @param hits - Ascending timestamps; mutated in place.
 * @param cutoff - Anything at or before this instant is expired.
 */
function pruneHits(hits: number[], cutoff: number): void {
  let keepFrom = 0;
  while (keepFrom < hits.length && hits[keepFrom] <= cutoff) keepFrom += 1;
  if (keepFrom > 0) hits.splice(0, keepFrom);
}

/**
 * Seconds a caller must wait for the oldest hit in a window to expire.
 *
 * Reported rather than a fixed guess, so a caller that honours `Retry-After` comes back exactly
 * when it can succeed instead of being told to wait a round number and failing again.
 *
 * @param hits - Ascending timestamps inside the window.
 * @param windowMs - Width of the window.
 * @param now - Current instant.
 * @returns Whole seconds to wait, at least one.
 */
function retryAfterSeconds(hits: number[], windowMs: number, now: number): number {
  const oldest = hits[0];
  if (oldest === undefined) return 1;
  return Math.max(1, Math.ceil((oldest + windowMs - now) / SECOND));
}

/**
 * Sliding-window limiter holding its counters in process memory.
 *
 * Instantiated once per module load; see the file header for what that does and does not
 * guarantee in a serverless environment.
 */
export class RateLimiter {
  private readonly config: RateLimitConfig;
  private readonly clients = new Map<string, ClientRecord>();
  private budget: BudgetEntry[] = [];

  /**
   * @param config - Policy to enforce. Defaults to {@link DEFAULT_RATE_LIMIT}.
   */
  constructor(config: RateLimitConfig = DEFAULT_RATE_LIMIT) {
    this.config = config;
  }

  /**
   * Decide whether a request may proceed, recording it if so.
   *
   * The shared budget is checked here but charged in {@link recordUsage}, once the real cost is
   * known. A request admitted on the last unit of budget is therefore allowed to finish; the
   * budget goes negative-ish by at most one question, which is the right trade against refusing
   * a caller for a cost that had not been incurred yet.
   *
   * @param clientId - Stable identifier for the caller, from {@link clientIdFromHeaders}.
   * @param now - Current instant in milliseconds; injectable so the policy is testable.
   * @returns Whether the request may proceed, and if not, why and for how long.
   */
  check(clientId: string, now: number = Date.now()): RateLimitDecision {
    this.pruneBudget(now);

    const spent = this.budget.reduce((total, entry) => total + entry.cost, 0);
    if (spent >= this.config.budgetLimit) {
      const oldest = this.budget[0]?.at ?? now;
      return {
        allowed: false,
        code: "daily_budget_exhausted",
        message:
          "This deployment has used its model budget for today. The tools themselves are " +
          "unaffected -- every /api endpoint below the agent keeps working, and the agent " +
          "returns tomorrow. Running locally with your own provider key has no such limit.",
        retryAfterS: Math.max(1, Math.ceil((oldest + this.config.budgetWindowMs - now) / SECOND)),
      };
    }

    const record = this.clients.get(clientId) ?? { hits: [] };
    pruneHits(record.hits, now - this.config.sustainedWindowMs);

    const burstHits = record.hits.filter((at) => at > now - this.config.burstWindowMs);
    if (burstHits.length >= this.config.burstLimit) {
      // Store the pruned record even on refusal, so a caller hammering the endpoint cannot use
      // rejected requests to keep its own window from being cleaned up.
      this.clients.set(clientId, record);
      return {
        allowed: false,
        code: "rate_limited",
        message: `Too many requests. This endpoint allows ${this.config.burstLimit} questions per minute.`,
        retryAfterS: retryAfterSeconds(burstHits, this.config.burstWindowMs, now),
      };
    }

    if (record.hits.length >= this.config.sustainedLimit) {
      this.clients.set(clientId, record);
      return {
        allowed: false,
        code: "rate_limited",
        message: `Too many requests. This endpoint allows ${this.config.sustainedLimit} questions per hour.`,
        retryAfterS: retryAfterSeconds(record.hits, this.config.sustainedWindowMs, now),
      };
    }

    record.hits.push(now);
    this.clients.set(clientId, record);
    this.evictIfCrowded();

    return { allowed: true };
  }

  /**
   * Charge the shared budget for work that actually happened.
   *
   * @param modelCalls - Model calls the request consumed. Floored at one, because a request that
   *   reached the provider cost something even if it failed before reporting a step.
   * @param now - Current instant in milliseconds.
   */
  recordUsage(modelCalls: number, now: number = Date.now()): void {
    this.budget.push({ at: now, cost: Math.max(1, Math.round(modelCalls)) });
    this.pruneBudget(now);
  }

  /** Model calls charged inside the current budget window. Exposed for tests and diagnostics. */
  spent(now: number = Date.now()): number {
    this.pruneBudget(now);
    return this.budget.reduce((total, entry) => total + entry.cost, 0);
  }

  /** How many clients are currently held in memory. Exposed so the bound can be asserted. */
  trackedClients(): number {
    return this.clients.size;
  }

  /** Drop budget charges that have aged out of the window. */
  private pruneBudget(now: number): void {
    const cutoff = now - this.config.budgetWindowMs;
    let keepFrom = 0;
    while (keepFrom < this.budget.length && this.budget[keepFrom].at <= cutoff) keepFrom += 1;
    if (keepFrom > 0) this.budget = this.budget.slice(keepFrom);
  }

  /**
   * Bound memory by discarding the least recently seen clients.
   *
   * Expiry alone is not enough: a flood of distinct identifiers can add entries faster than the
   * sustained window retires them, and an unbounded map in a long-lived instance is a leak.
   */
  private evictIfCrowded(): void {
    if (this.clients.size <= this.config.maxTrackedClients) return;

    const byLastSeen = [...this.clients.entries()].sort(
      (a, b) => (a[1].hits.at(-1) ?? 0) - (b[1].hits.at(-1) ?? 0),
    );

    const excess = this.clients.size - this.config.maxTrackedClients;
    for (let index = 0; index < excess; index += 1) {
      this.clients.delete(byLastSeen[index][0]);
    }
  }
}

/**
 * Identify the caller from proxy headers.
 *
 * Vercel sets `x-forwarded-for` on every request, with the client address first, and **it
 * overwrites whatever the caller sent**. Measured against the deployment rather than assumed:
 * 130 requests each claiming a different address were limited at exactly 120, the per-minute
 * allowance, instead of passing as 130 separate clients. So identity here cannot be rotated by
 * spoofing the header.
 *
 * It is still not proof of identity -- a real botnet has real addresses, and a NAT puts many
 * people behind one. That is what the global budget is for: it holds regardless of how many
 * distinct addresses a caller actually has.
 *
 * @param headers - Request headers.
 * @returns A stable key for the caller, or a shared fallback when no address is present.
 */
export function clientIdFromHeaders(headers: Headers): string {
  const forwarded = headers.get("x-forwarded-for");
  if (forwarded) {
    const first = forwarded.split(",")[0]?.trim();
    if (first) return first;
  }

  const real = headers.get("x-real-ip")?.trim();
  if (real) return real;

  // Everything unidentifiable shares one bucket. Local development has no proxy header, which is
  // the usual way to land here.
  return "unidentified";
}

/**
 * Whether limits apply in this environment.
 *
 * Enforced only on a deployment. The evaluation harness fires 157 questions at a local server in
 * one run, and a limiter that blocked that would be measuring itself instead of the agent. The
 * policy is still verified in development -- by its own tests, which drive {@link RateLimiter}
 * directly rather than through the route.
 *
 * @returns True when the process is a production build.
 */
export function rateLimitingEnabled(): boolean {
  return process.env.NODE_ENV === "production";
}

/**
 * Process-wide limiter for the agent route.
 *
 * Module scope is deliberate: it must survive between requests handled by the same instance,
 * which is exactly what makes it effective under Fluid Compute's instance reuse.
 */
export const agentRateLimiter = new RateLimiter();
